gammapy/spectrum/flux_point.py

```diff
@@ -1041,68 +1041,87 @@ class FluxPointFit(Fit):
         print(result)
         print(result.model)
     """
 
-    def __init__(self, model, data, stat="chi2"):
-        self._model = model
+    def __init__(self, model, data, mask=None, likelihood="chi2"):
+        self.model = model
         self.data = data
+        self.mask = mask
+        self.parameters = model.parameters
 
-        if stat in ["chi2", "chi2assym"]:
-            self._stat = stat
+        if likelihood in ["chi2", "chi2assym"]:
+            self._likelihood = likelihood
         else:
             raise ValueError(
-                "'{stat}' is not a valid fit statistic, please choose"
+                "'{likelihood}' is not a valid fit statistic, please choose"
                 " either 'chi2' or 'chi2assym'"
             )
 
-    @property
-    def _stat_chi2(self):
-        """Likelihood per bin given the current model parameters"""
-        model = self._model(self.data.e_ref)
-        data = self.data.table["dnde"].quantity
-        sigma = self.data.table["dnde_err"].quantity
+    @staticmethod
+    def _likelihood_chi2(data, model, sigma):
         return ((data - model) / sigma).to_value("") ** 2
 
-    @property
-    def _stat_chi2_assym(self):
+    @staticmethod
+    def _likelihood_chi2_assym(data, model, sigma_n, sigma_p):
         """
         Assymetric chi2 statistics for a list of flux points and model.
         """
-        model = self._model(self.data.e_ref)
-        data = self.data.table["dnde"].quantity.to(model.unit)
-        data_errp = self.data.table["dnde_errp"].quantity
-        data_errn = self.data.table["dnde_errn"].quantity
+        is_p = (model > data)
+        sigma = sigma_n
+        sigma[is_p] = sigma_p[is_p]
+        return FluxPointsDataset._likelihood_chi2(data, model, sigma)
 
-        val_n = ((data - model) / data_errn).to_value("") ** 2
-        val_p = ((data - model) / data_errp).to_value("") ** 2
-        return np.where(model > data, val_p, val_n)
+    def flux_pred(self):
+        """Compute predicted flux.
+        """
+        return self.model(self.data.e_ref)
 
-    @property
-    def stat(self):
-        if self._stat == "chi2":
-            return self._stat_chi2
+    def likelihood_per_bin(self):
+        """Likelihood per bin given the current model parameters"""
+        model = self.flux_pred()
+        data = self.data.table["dnde"].quantity
+
+        if self._likelihood == "chi2":
+            sigma = self.data.table["dnde_err"].quantity
+            return self._likelihood_chi2(data, model, sigma)
+        elif self._likelihood == "chi2assym":
+            sigma_n = self.data.table["dnde_errn"].quantity
+            sigma_p = self.data.table["dnde_errp"].quantity
+            return self._likelihood_chi2_assym(data, model, sigma_n, sigma_p)
         else:
-            return self._stat_chi2_assym
+            # TODO: add likelihood profiles
+            pass
 
-    def total_stat(self, parameters):
+    def likelihood(self, parameters):
```

Comment:
Hello @adonath ! I was wondering why you are passing `parameters` here?

Reply:
That's a left over from the previous implementation of `total_stat()`. It could be just removed here and in `SpectrumFit.total_stat` as well...